engine/src/main/java/org/hibernate/validator/internal/engine/constraintvalidation/ConstraintValidatorManager.java

```diff
@@ -94,87 +94,52 @@ public ConstraintValidatorManager(ConstraintValidatorFactory constraintValidator
-				constraintFactory,
-				validatorDescriptor,
-				descriptor
-		);
 		if ( constraintValidator == null ) {
-			putInitializedValidator(
-					validatedValueType,
-					descriptor.getAnnotation(),
-					constraintFactory,
-					DUMMY_CONSTRAINT_VALIDATOR
-			);
-			return null;
+			constraintValidator = createAndInitializeValidator( validatedValueType, descriptor, constraintFactory );
+			constraintValidator = cacheValidator( key, constraintValidator );
 		}
 		else {
-			putInitializedValidator(
-					validatedValueType,
-					descriptor.getAnnotation(),
-					constraintFactory,
-					constraintValidator
-			);
-			return constraintValidator;
-
+			LOG.tracef( "Constraint validator %s found in cache.", constraintValidator );
 		}
+
+		return DUMMY_CONSTRAINT_VALIDATOR == constraintValidator ? null : constraintValidator;
 	}
 
-	private void putInitializedValidator(Type validatedValueType,
-			Annotation annotation,
-			ConstraintValidatorFactory constraintFactory,
-			ConstraintValidator<?, ?> constraintValidator) {
-		// we only cache constraint validator instance for the default and least recently used factory
-		if ( constraintFactory != defaultConstraintValidatorFactory && constraintFactory != mostRecentlyUsedNonDefaultConstraintValidatorFactory ) {
+	private <A extends Annotation> ConstraintValidator<A, ?> cacheValidator(CacheKey key,
+			ConstraintValidator<A, ?> constraintValidator) {
+		// we only cache constraint validator instances for the default and most recently used factory
+		if ( key.constraintFactory != defaultConstraintValidatorFactory && key.constraintFactory != mostRecentlyUsedNonDefaultConstraintValidatorFactory ) {
 			clearEntriesForFactory( mostRecentlyUsedNonDefaultConstraintValidatorFactory );
-			mostRecentlyUsedNonDefaultConstraintValidatorFactory = constraintFactory;
+			mostRecentlyUsedNonDefaultConstraintValidatorFactory = key.constraintFactory;
 		}
 
-		final CacheKey key = new CacheKey(
-				annotation,
-				validatedValueType,
-				constraintFactory
-		);
+		@SuppressWarnings("unchecked")
+		ConstraintValidator<A, ?> cached = (ConstraintValidator<A, ?>) constraintValidatorCache.putIfAbsent( key, constraintValidator );
 
-		constraintValidatorCache.putIfAbsent( key, constraintValidator );
+		return cached != null ? cached : constraintValidator;
 	}
 
-	private <A extends Annotation> ConstraintValidator<A, ?> createAndInitializeValidator(
-			ConstraintValidatorFactory constraintFactory,
-			ConstraintValidatorDescriptor<A> validatorDescriptor,
-			ConstraintDescriptor<A> descriptor) {
+	@SuppressWarnings("unchecked")
+	private <A extends Annotation> ConstraintValidator<A, ?> createAndInitializeValidator(Type validatedValueType,
+			ConstraintDescriptorImpl<A> descriptor,
+			ConstraintValidatorFactory constraintFactory) {
+
+		ConstraintValidatorDescriptor<A> validatorDescriptor = findMatchingValidatorDescriptor( descriptor, validatedValueType );
+		ConstraintValidator<A, ?> constraintValidator = null;
```

Comment:
We shouldn't initialize `constraintValidator` here, it's useless. I'll change that and merge.